sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/exemplar/AbstractFixedSizeExemplarReservoir.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.exemplar;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.common.Clock;
+import io.opentelemetry.sdk.metrics.data.DoubleExemplar;
+import io.opentelemetry.sdk.metrics.data.Exemplar;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Base implementation for fixed-size reservoir sampling of Exemplars.
+ *
+ * <p>Additionally this implementation ONLY exports double valued exemplars.
+ */
+abstract class AbstractFixedSizeExemplarReservoir implements ExemplarReservoir {
+  private final Clock clock;
+  private final ReservoirCell[] storage;
+
+  /**
+   * Instantiates an exemplar reservoir of fixed size.
+   *
+   * @param clock The clock to use when annotating measurements with time.
+   * @param size The number of exemplars to preserve.
+   */
+  AbstractFixedSizeExemplarReservoir(Clock clock, int size) {
+    this.clock = clock;
+    this.storage = new ReservoirCell[size];
+    for (int i = 0; i < size; ++i) {
+      this.storage[i] = new ReservoirCell();
+    }
+  }
+
+  protected final int maxSize() {
+    return storage.length;
+  }
+  /**
+   * Determines the sample bucket for a given measurement.
+   *
+   * @return The bucket to sample into or -1 for no sampling.
+   */
+  protected abstract int bucketFor(double value, Attributes attributes, Context context);
```

Comment:
I wonder if this method shouldn't be "reservoirIndexFor", since we don't really call things "buckets" unless it's a histogram, I don't think.